src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,241 @@
+    const float gain = pt2FilterGain(cutoffHz, taskIntervalSeconds);
+    pt2FilterInit(&altHoldDeltaLpf, gain); 
+
+    altHoldState.isAltHoldActive = false;
+    altHoldReset();
+}
+
+void altHoldProcessTransitions(void) {
+
+    if (FLIGHT_MODE(ALTHOLD_MODE)) {
+        if (!altHoldState.isAltHoldActive && isAltitudeAvailable()) {
+            altHoldReset();
+            altHoldState.isAltHoldActive = true;
+        }
+    } else {
+        altHoldState.isAltHoldActive = false;
+    }
+
+    // ** the transition out of alt hold (exiting altHold) may be rough.  Some notes... **
+    // The original PR had a gradual transition from hold throttle to pilot control throttle
+    // using !(altHoldRequested && altHoldState.isAltHoldActive) to run an exit function
+    // a cross-fade factor was sent to mixer.c based on time since the flight mode request was terminated
+    // it was removed primarily to simplify this PR
+
+    // hence in this PR's the user's throttle needs to be close to the hover throttle value on exiting altHold
+    // its not so bad because the 'target adjustment' by throttle requires that
+    // user throttle must be not more than half way out from hover for a stable hold
+}
+
+void altHoldUpdateTargetAltitude(void)
+{
+    // The user van raise or lower the target altitude with throttle up;  there is a big deadband.
+    // Max rate for climb/descend is 1m/s by default (up to 2.5 is allowed but overshoots a fair bit)
+    // If set to zero, the throttle has no effect.
+
+    // Some people may not like throttle being able to change the target altitude, because:
+    // - with throttle adjustment, hitting the switch won't always hold altitude if throttle is bumped
+    // - eg if the throttle is bumped low accidentally, quad will start descending.
+    // On the plus side:
+    // - the pilot has control nice control over altitude, and the deadband is wide
+    // - Slow controlled descents are possible, eg for landing
+    // - fine-tuning height is possible, eg if there is slow sensor drift
+    // - to keep the craft stable, throttle must be in the deadband, making exits smoother
+
+    const float rcThrottle = rcCommand[THROTTLE];
+
+    const float lowThreshold = 0.5f * (positionConfig()->hover_throttle + PWM_RANGE_MIN); // around 1150
+    const float highThreshold = 0.5f * (positionConfig()->hover_throttle + PWM_RANGE_MAX); // around 1500
+    
+    float throttleAdjustmentFactor = 0.0f;
+    if (rcThrottle < lowThreshold) {
+        throttleAdjustmentFactor = scaleRangef(rcThrottle, PWM_RANGE_MIN, lowThreshold, -1.0f, 0.0f);
+    } else if (rcThrottle > highThreshold) {
+        throttleAdjustmentFactor = scaleRangef(rcThrottle, highThreshold, PWM_RANGE_MAX, 0.0f, 1.0f);
+    }
+
+    // if failsafe is active, and we get here, we are in landing mode, so descend at configured max rate
+    if (failsafeIsActive()) {
+        throttleAdjustmentFactor = -1.0f;
```

Comment:
Is it desirable to land at maximum configured speed?

Reply:
In the 'emergency' or 'failsafe' Landing Mode, there is a time out before the quad will just disarm and crash.  The default is 10s and max is 200s.  That default of 10s means an inevitable in-air disarm if the landing mode is initiated at a height of more than 10m.
Maybe with this code the time default should be 100s now, to be OK even if 100m up.  
Previously it was just 10s because landing mode was useless, in any practical sense.
In practice a landing at 1m/s is very unlikely to be destructive.
The auto-disarm feature will disarm more reliably if the impact is not very gentle.  At 1m/s on soft grass the impact is very gentle and doesn't always disarm with threshold of 50.
In the last 20cm of descent, there is an air-cushion under the quad that softens the impact.  The air pressure increase is detected by Baro and leads to an apparent drop in altitude, leading to a PID driven increase in motors, further softening the landing (and tending to make it bouncy).
So yes 1m/s is OK.
In GPS Rescue Descent mode, I increased the descent rate when altitudes were high, up to 3x the set value, otherwise you had to wait a long time.  Should I consider doing that here?

Reply:
OK I increased descent rate in failsafe landing mode by factor of 2 when altitude is above landing altitude.